builtin/k8s: Implement `WatchTask` for k8s task plugin

briancain opened the pull request:
This commit introduces an implementation of WatchTask for the kubernetes
Task plugin. It will poll initially to ensure that the pod being watched
properly starts up before attempting to stream its logs. Once the pod
has started, WatchTask will stream the pod logs and send it to the
runner logs as well as the UI output for the job.

briancain committed:
builtin/k8s: Implement `WatchTask` for k8s task plugin

This commit introduces an implementation of WatchTask for the kubernetes
Task plugin. It will poll initially to ensure that the pod being watched
properly starts up before attempting to stream its logs. Once the pod
has started, WatchTask will stream the pod logs and send it to the
runner logs as well as the UI output for the job.

paladin-devops reviewed (commented):
Looks great! Just had one suggestion for the future regarding timeouts, but definitely can wait for a later PR.

briancain commented on builtin/k8s/task.go:
```diff
@@ -402,9 +407,125 @@ func (p *TaskLauncher) StartTask(
+		}
+		// NOTE(briancain): it doesn't seem like the k8s API is sending an EOF??
+		// This loop never exits here
+		if err == io.EOF {
```
That's a better way to go about this for sure! Will update this PR to check for that instead of waiting for no output from the logs.

briancain replied:
This should be resolved via 952fad8b256a6af419e2b1b0048bbe77adb80ec1 !

briancain commented on builtin/k8s/task.go:
```diff
@@ -402,9 +407,125 @@ func (p *TaskLauncher) StartTask(
+	pod := pods.Items[0]
+
+	// TODO(briancain): a better wait time duration?
+	timeout := time.Duration(30 * time.Second)
```
I think this timeout might go away if I add Mitchells suggestion for a channel that watches the pod status and exits that way. But agreed configurable timeouts would be good, I will see if this is still needed after adding the pod status watcher!

briancain commented on builtin/k8s/task.go:
```diff
@@ -402,9 +407,125 @@ func (p *TaskLauncher) StartTask(
+
+			continue
+		}
+		// NOTE(briancain): it doesn't seem like the k8s API is sending an EOF??
```
That makes sense, and would explain why I wasn't seeing EOFs! If that's the case, all the more reason to have a channel watcher on the pod status here over waiting for an EOF.

briancain committed:
k8s/task: Add configurable timeout for waiting for pod to start up

briancain committed:
k8s/task: Watch pod phase for finishing WatchTask log stream

This commit updates the WatchTask k8s plugin to spawn a go func that
inspects the pod phase while waiting for any logstream bytes from the
pod. If the pod has wrapped up and there's no more data coming from the
log stream, WatchTask will exit.

briancain committed:
lint: Fix go linter for ci test

briancain committed:
Remove extra message in log

Only log message from runner.

briancain committed:
k8s/task: Set task exit code based on pod phase

briancain commented:
Rebased this off main to include the `WatchTask` job additions. I also included the commits from https://github.com/hashicorp/waypoint/pull/3423 and will rebase those out once that PR is merged. But wanted to include it for now so that this PR keeps working.

briancain commented on builtin/k8s/task.go:
```diff
@@ -402,9 +421,160 @@ func (p *TaskLauncher) StartTask(
+			case v1.PodPending, v1.PodUnknown:
+				// Unknown state, still wait
+				logsDoneCh <- false
+			}
```
We can put a sleep here and exit on context cancel so that we're not hammering the k8s API looking at the pod phase.

briancain commented on builtin/k8s/task.go:
```diff
@@ -402,9 +421,160 @@ func (p *TaskLauncher) StartTask(
+			break
+		}
+		if err != nil {
+			log.Error("got an error streaming pod logs", "err", err)
```
```golang
ui.Output("error streaming logs: %s", err, terminal.WithErrorStyle())
```

briancain replied:
We could stream the error to the WatchTask jobstream like this else where.

briancain commented on builtin/k8s/task.go:
```diff
@@ -402,9 +421,160 @@ func (p *TaskLauncher) StartTask(
+			// hasn't sent any log bytes.
+			select {
+			case <-logsDoneCh:
+				log.Trace("pod is finished")
```
We should also log the pod id here and maybe improve the message here.

briancain committed:
Add more output to WatchTask job stream, include pod name

briancain committed:
k8s/task: Sleep a bit when polling for pod phase for logs

This commit updates the go func watcher that watches a pod phase while
log streaming to sleep a bit so it doesn't hammer the k8s cluster
inspecting the pod phase.
